controller/controller.go

```diff
@@ -0,0 +1,147 @@
+package controller
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"github.com/dgrijalva/jwt-go"
+	"iotdashboard/dbmanager"
+	"golang.org/x/crypto/bcrypt"
+	"time"
+)
+
+//TODO: these probably shouldn't be global vars
+var PSQL dbmanager.DBManager
+var jwtKey []byte
+
+// key is the JWT, value is the expiration epoch time
+var blocklist map[string]int64
+
+type Claims struct {
+	NotBefore int64 `json:"nbf"`
+	jwt.StandardClaims
+}
+
+////////Errors
+var ErrExpiredToken = errors.New("Token is expired")
+
+func init() {
+	var err error
+	jwtKey, err = GenerateRandomToken(256)
+	if err != nil {
+		panic(err)
+	}
+	PSQL, err = dbmanager.New("postgres", "myPassword", "iot_dashboard")
+	if err != nil {
+		panic(err)
+	}
+
+	blocklist = make(map[string]int64)
+}
+
+func Login(email, password string) (string, error) {
+	// use user struct instead of email and password?
+
+	// validate CSRF
+
+	// validate basic auth
+	err := CheckUserCredentials(email, password)
+	if err != nil {
+		return "", err
+	}
+
+	// create and return JWT
+	var token string
+	token, err = CreateJWT(60)
+	if err != nil {
+		return "", err
+	}
+	return token, nil
+
+}
+
+func Logout(token string) error {
+	// validate CSRF
+
+	// validate JWT
+	exp, err := ValidateJWT(token)
+	if err != nil {
+		return err
+	}
+	// blocklist JWT
+	blocklist[token] = exp
+	return nil
+
+}
+
+// returns an Error if the credentials are not valid
+func CheckUserCredentials(email, password string) error {
+	hash, err := PSQL.GetUserPassHash(email)
+	if err != nil {
+		return err
+	}
+	return bcrypt.CompareHashAndPassword(hash, []byte(password))
+}
+
+func CreateJWT(validPeriod int64) (string, error) {
+	claims := Claims{
+		NotBefore: time.Now().Unix(),
+		StandardClaims: jwt.StandardClaims{
+			// In JWT, the expiry time is expressed as unix milliseconds
+			ExpiresAt: time.Now().Add(time.Second * time.Duration(validPeriod)).Unix(),
```

Comment:
It seems like you forgot to add `UTC()` in there

Reply:
@pierrebeaucamp doesn't Unix() return the number of seconds in epoch time, which is UTC?

Reply:
Yes, you're right 🙂, I should have read that more carefully. I was mostly reading back and forth and saw that you did `.UTC()` on Line 126:

```golang
now := time.Now().UTC().Unix()
```

Personally, I got into the habit of adding `.UTC()` to each `time.Now()` call so the logic is correct wether or not you later call `Unix()` on the result. But it's up to you, I would just aim to keep things consistent.

Reply:
I added .UTC() to these calls for readability